src/NHibernate/AdoNet/DbBatchBatcher.cs

```diff
@@ -0,0 +1,297 @@
+			}
+		}
+
+		private void AddCommandToBatch(DbCommand batchUpdate)
+		{
+			var dbBatchCommand = _currentBatch.CreateBatchCommand();
+			dbBatchCommand.CommandText = batchUpdate.CommandText;
+			dbBatchCommand.CommandType = batchUpdate.CommandType;
+
+			foreach (var param in batchUpdate.Parameters)
+			{
+				dbBatchCommand.Parameters.Add(((ICloneable) param).Clone());
+			}
+			_currentBatch.BatchCommands.Add(dbBatchCommand);
+		}
+
+		public override Task AddToBatchAsync(IExpectation expectation, CancellationToken cancellationToken)
+		{
+			if (cancellationToken.IsCancellationRequested)
+			{
+				return Task.FromCanceled<object>(cancellationToken);
+			}
+			try
+			{
+				_totalExpectedRowsAffected += expectation.ExpectedRowCount;
+				var batchUpdate = CurrentCommand;
+				Driver.AdjustCommand(batchUpdate);
+				string lineWithParameters = null;
+				var sqlStatementLogger = Factory.Settings.SqlStatementLogger;
+				if (sqlStatementLogger.IsDebugEnabled || Log.IsDebugEnabled())
+				{
+					lineWithParameters = sqlStatementLogger.GetCommandLineWithParameters(batchUpdate);
+					var formatStyle = sqlStatementLogger.DetermineActualStyle(FormatStyle.Basic);
+					lineWithParameters = formatStyle.Formatter.Format(lineWithParameters);
+					_currentBatchCommandsLog.Append("command ")
+											.Append(_currentBatch.BatchCommands.Count)
+											.Append(":")
+											.AppendLine(lineWithParameters);
+				}
+				if (Log.IsDebugEnabled())
+				{
+					Log.Debug("Adding to batch:{0}", lineWithParameters);
+				}
+
+				AddCommandToBatch(batchUpdate);
+
+				if (CountOfStatementsInCurrentBatch >= _batchSize)
+				{
+					return ExecuteBatchWithTimingAsync(batchUpdate, cancellationToken);
+				}
+				return Task.CompletedTask;
+			}
+			catch (Exception ex)
+			{
+				return Task.FromException<object>(ex);
+			}
+		}
+
+		protected override void DoExecuteBatch(DbCommand ps)
+		{
+			try
+			{
+				Log.Debug("Executing batch");
+				CheckReaders();
+				Prepare(_currentBatch);
+				if (Factory.Settings.SqlStatementLogger.IsDebugEnabled)
+				{
+					Factory.Settings.SqlStatementLogger.LogBatchCommand(_currentBatchCommandsLog.ToString());
+				}
+				int rowsAffected;
+				try
+				{
+					rowsAffected = _currentBatch.ExecuteNonQuery();
+				}
+				catch (DbException e)
+				{
+					throw ADOExceptionHelper.Convert(Factory.SQLExceptionConverter, e, "could not execute batch command.");
+				}
+				
+				Expectations.VerifyOutcomeBatched(_totalExpectedRowsAffected, rowsAffected, ps);
+			}
+			finally
+			{
+				ClearCurrentBatch();
+			}
+		}
+
+		protected override async Task DoExecuteBatchAsync(DbCommand ps, CancellationToken cancellationToken)
+		{
+			cancellationToken.ThrowIfCancellationRequested();
+			try
+			{
+				Log.Debug("Executing batch");
+				await (CheckReadersAsync(cancellationToken)).ConfigureAwait(false);
+				//await (PrepareAsync(_currentBatch, cancellationToken)).ConfigureAwait(false);
```

Comment:
Why not prepare async?

Reply:
Good catch. I think I was experimenting with getting the AsyncGenerator to do the job, but got stuck and forgot this one. I'm guessing the compiler directive can get in the way of AsyncGenerator?

Fixed, manually.